### PR TITLE
Use correct product version in SPC in servicing

### PR DESCRIPTION
**Pull request by akoeplinger**
I noticed we missed to port the version changes from https://github.com/dotnet/runtime/pull/108899#pullrequestreview-2371679935 to main. This does that so we embed the correct version number for `FrameworkDescription` into System.Private.CoreLib even during servicing when using the stabilized version.

I simplified the source generator to pass the InformationalVersion msbuild property into the compilation instead of using reflection on the generator assembly.

### CODE REVIEW

File: src/libraries/System.Runtime/tests/System.Runtime.InteropServices.RuntimeInformation.Tests/DescriptionNameTests.cs [hunk 1]
```diff
@@ -196,17 +196,10 @@ public void VerifyFrameworkDescriptionContainsCorrectVersion()
                 return;
 
             Assert.DoesNotContain("+", version); // no git hash
+            Assert.DoesNotContain("-", version); // no preview suffix
```
Comment (jkotas) [inline src/libraries/System.Runtime/tests/System.Runtime.InteropServices.RuntimeInformation.Tests/DescriptionNameTests.cs:199]: It was possible to identify preview or private builds from the FrameworkDescription string before this change. This feature has been useful for diagnostics. Should this feature be preserved?

Comment (akoeplinger) [reply src/libraries/System.Runtime/tests/System.Runtime.InteropServices.RuntimeInformation.Tests/DescriptionNameTests.cs]: @jkotas yeah, we seem to have other tests that make that assumption so it was easier to restore the earlier behavior.

**Review by ViktorHofer — APPROVED** [review]

File: src/libraries/System.Private.CoreLib/gen/ProductVersionInfoGenerator.cs [hunk 2]
```diff
@@ -12,18 +12,25 @@ public partial class ProductVersionInfoGenerator : IIncrementalGenerator
     {
         public void Initialize(IncrementalGeneratorInitializationContext context)
         {
-            context.RegisterPostInitializationOutput(ctx =>
-            {
-                string? informationalVersion = typeof(ProductVersionInfoGenerator).Assembly.GetCustomAttribute<AssemblyInformationalVersionAttribute>()?.InformationalVersion;
+            var informationalVersionProvider = context.AnalyzerConfigOptionsProvider.Select((options, _) =>
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/gen/ProductVersionInfoGenerator.cs:15]:
```suggestion
            string informationalVersionProvider = context.AnalyzerConfigOptionsProvider.Select((options, _) =>
```
Nit: Repo coding conventions

Comment (akoeplinger) [reply src/libraries/System.Private.CoreLib/gen/ProductVersionInfoGenerator.cs:15]: This is not a string but Roslyn's `IncrementalValueProvider<T>`: https://github.com/dotnet/roslyn/blob/main/docs%2Ffeatures%2Fincremental-generators.md#incrementalvaluesprovidert

Comment (akoeplinger) [reply src/libraries/System.Private.CoreLib/gen/ProductVersionInfoGenerator.cs:15]: ah sorry, I missed that this was about the coding conventions: https://github.com/dotnet/runtime/pull/110997

**Pull request merged**